### PR TITLE
Apply permutation to state-vector with decomposition based synthesis

### PR DESCRIPTION
**Pull request by msoeken**
This operation applies a permutation to a state-vector using a technique called decomposition based synthesis explained in the paper [[A. De Vos and Y. Van Rentergem, Young subgroups for reversible computers, *Adv. in Math. of Comm.* **2** (2008), 183-200](http://www.aimsciences.org/article/doi/10.3934/amc.2008.2.183)]

It computes a sequence of Boolean functions that can be used as control functions to control X gates using the API proposed in #271.

TODO list:
* [x] Unify tests with `ApplyPermutationUsingTransformation`
* [x] Make `Adj + Ctl` possible
* [x] Docs
* [x] Give the possibility to define a decomposition order
* [x] Check pre-conditions

*Note:* #271 should be merged before this one, then the file changes will be subsumed.

### CODE REVIEW

File: Standard/src/Synthesis/DecompositionBased.qs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+        return ((value &&& ~~~mask) <<< 1) + (value &&& mask);
+    }
+
+    /// # Summary
+    /// State during decomposition based on variable indexes
+    ///
+    /// # Description
+    /// The state holds the current permutation and the currently generated functions
+    /// for controlled gates on the left, and controlled gates on the right.
+    internal newtype DecompositionState = (perm : Int[], lfunctions : (BigInt, Int)[], rfunctions : (BigInt, Int)[]);
```
Comment (cgranade) [inline Standard/src/Synthesis/DecompositionBased.qs:85]: In general, named items for UDTs start with upper-case letters.

File: Standard/src/Synthesis/DecompositionBased.qs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    /// ## perm
+    /// A permutation of $2^n$ elements starting from 0.
+    /// ## qubits
+    /// A list of $n$ qubits to which the permutation is applied to.
+    ///
+    /// # Example
+    /// To synthesize a `SWAP` operation:
+    /// ```Q#
+    /// using (qubits = Qubit[2]) {
+    ///   ApplyPermutationUsingDecomposition([0, 2, 1, 3], qubits);
```
Comment (cgranade) [inline Standard/src/Synthesis/DecompositionBased.qs:157]: Similarly here, should this be `LittleEndian(qubits)`?

File: Standard/src/Synthesis/DecompositionBased.qs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+
+    /// # Summary
+    /// Decomposes a permutation on a variable
+    ///
+    /// # Description
+    /// Given a permutation $\pi$ (`perm`) and an index $i$ (`index`), this method
+    /// returns three permutations $((\pi_l, \pi_r), \pi')$ such that the images
+    /// of $\pi_l$ and $\pi_r$ do not change bits in their elements at indexes other
+    /// than $i$ and images of $\pi'$ do not change bit $i$ in their elements.
+    internal function DecomposedOn(perm : Int[], index : Int) : ((Int[], Int[]), Int[]) {
```
[the remark this answers is not in the document]
Comment (msoeken) [reply Standard/src/Synthesis/DecompositionBased.qs:25]: This is really tricky to explain. If I just show the resulting permutations, it's not obvious at all where they come from. But I will give it a try with a more elaborated explanation.

[the remark this answers is not in the document]
Comment (msoeken) [reply Standard/src/Synthesis/DecompositionBased.qs:92]: I added some more explanation.

File: Standard/src/Synthesis/DecompositionBased.qs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+        ApplyPermutationUsingDecompositionWithVariableOrder(perm, SequenceI(0, Length(qubits!) - 1), qubits);
+    }
+
+    /// # Summary
+    /// Permutes the amplitudes in a quantum state given a permutation
+    /// using decomposition-based synthesis.
+    ///
+    /// # Description
+    /// This operation is a more general version of @"microsoft.quantum.synthesis.applypermutationusingdecomposition"
+    /// in which the variable order can be specified.
```
[the remark this answers is not in the document]
Comment (msoeken) [reply Standard/src/Synthesis/DecompositionBased.qs]: I explained that the change of `variableOrder` leads to different number of gates.

**Pull request merged**